sources/api/pluto/src/main.rs

```diff
@@ -144,42 +162,37 @@ async fn get_max_pods(client: &mut ImdsClient) -> Result<String> {
     error::NoInstanceTypeMaxPods { instance_type }.fail()
 }
 
-/// Returns the cluster's DNS IPV4 address. First it attempts to call EKS describe-cluster to find
-/// the `serviceIPv4CIDR`. If that works, it returns the expected cluster DNS IP address which is
-/// obtained by substituting `10` for the last octet. If the EKS call is not successful, it falls
-/// back to using IMDS MAC CIDR blocks to return one of two default addresses.
-async fn get_cluster_dns_ip(client: &mut ImdsClient) -> Result<String> {
-    // try calling eks describe-cluster to figure out the dns cluster ip
-    if let Some(dns_ip) = get_dns_from_eks().await {
-        // we were able to calculate the dns ip from the cidr range we received from eks
-        return Ok(dns_ip);
+/// Returns the cluster's DNS address.
+///
+/// For IPv4 clusters, first it attempts to call EKS describe-cluster to find the `serviceIpv4Cidr`.
+/// If that works, it returns the expected cluster DNS IP address which is obtained by substituting
+/// `10` for the last octet. If the EKS call is not successful, it falls back to using IMDS MAC CIDR
+/// blocks to return one of two default addresses.
+///
+/// For IPv6 clusters, EKS currently is choosing a fixed unique local address (ULA) of `fdf9:2874:204b::/108`
+/// for cluster IPs. Cluster DNS IP in this case would always be hardcoded to `fdf9:2874:204b::a`.
+/// (The `::a` is equivalent to `.10` for IPv4)
+/// FIXME Update this once EKS changes their approach for determining the IPv6 service CIDR
+async fn get_cluster_dns_ip(
+    client: &mut ImdsClient,
+    network_config: KubernetesNetworkConfigResponse,
+) -> Result<String> {
+    if network_config.ip_family == Some("ipv6".to_string()) {
+        // For IPv6 clusters, `::a` is equivalent to `.10` for IPv4
+        return Ok(
+            DEFAULT_SERVICE_IPV6_CIDR.split('/').collect::<Vec<&str>>()[0].to_string() + "a",
+        );
     }
```

Comment:
Good idea. I'll try it out.

Reply:
Hmm, can you elaborate a bit more on why it's not good to have pluto's main.rs import `rusoto_eks`/EKS specific stuff?

`pluto` is already EKS specific (we only build `pluto` for aws-k8s variants) so I don't think it hurts to have it have EKS references in its code.

I tried the suggestion out a bit and started realizing I'm basically reconstructing the `rusoto_eks::KubernetesNetworkConfigResponse` struct in eks.rs just so main.rs wouldn't have to import something from `rusoto_eks` and it seems like a waste if I can just re-use that struct definition directly.

Reply:
I still think the `eks.rs` and `main.rs` separation make sense since we only make rusoto EKS API calls in `eks.rs`. But I don't think it's necessary to separate all EKS references in `main.rs` to `eks.rs` since that would almost be everything in `main.rs`.

Reply:
> basically reconstructing the rusoto_eks::KubernetesNetworkConfigResponse struct in eks.rs just so main.rs wouldn't have to import something from rusoto_eks and it seems like a waste if I can just re-use that struct definition directly.

Basically all of pluto is an implementation detail, so I wouldn't block on this. But my thinking for pushing code into eks.rs and creating our own `EksInfo` struct instead of using `rusoto_eks::KubernetesNetworkConfigResponse` is as follows:
- main.rs has a lot going on and it's honestly pretty hard to understand what's happening (it was before your changes and got worse). If you can a bunch of that code and put it somewhere else and call it with something like "get the dns cluster ip if possible", then half the cognitive load of reading main.rs is offloaded.
- Creating a small type doesn't cost anything, and if it returns exactly the data needed by a caller in exactly the way the caller needs it, then I think that's better than borrowing a (potentially bloated) type from a library.
- Changing from Rusoto to the new SDK would be isolated to eks.rs.
- `rusoto_eks::KubernetesNetworkConfigResponse` is a ridiculously long type name and I hate seeing it everywhere :)

> I don't think it's necessary to separate all EKS references in main.rs to eks.rs

I think of eks.rs more as rusoto encapsulation than EKS encapsulation.

Anyway, in summary, it's not a big deal. The main feedback is that main.rs is very dense and getting denser.

Reply:
> main.rs has a lot going on and it's honestly pretty hard to understand what's happening (it was before your changes and got worse). If you can a offload some code and put it elsewhere, and call it with something like "get the dns cluster ip if possible", then you offload some of the cognitive burdon in main.rs.

I think the issue is that all the logic here is tightly coupled so if we want to move `get_cluster_dns_ip` and `get_node_ip` to eks.rs then all the helpers/unit-tests/constants in main.rs will also move alongside, leaving main.rs with just the `run` and `main` function. So someone trying to understand what's going on `pluto` essentially would have to look at the functions, whether they be in `main.rs` or `eks.rs`. If they just want to understand at a high level of what's going on, I think the README at the top of main.rs and `run()` explains things pretty well.
And I don't particularly think the logic is that hard to understand with the docstrings and inline comments.


> Creating a small type doesn't cost anything, and if it returns exactly the data needed by a caller in exactly the way the caller needs it, then I think that's better than borrowing a (potentially bloated) type from a library.

Currently KubernetesNetworkConfigResponse contains exactly what we need, no more or less: `ipfamily`, `serviceipv4cidr`, `serviceipv6cidr`. If the type adds a bunch of type we don't need or if we need more information from other parts of the cluster struct down the road, we can always revisit and just create our own type.

> Changing from Rusoto to the new SDK would be isolated to eks.rs.
> KubernetesNetworkConfigResponse is a ridiculously long type name and I hate seeing it everywhere :)
I agree! I went ahead and aliased KubernetesNetworkConfigResponse as `ClusterNetworkInfo` in eks.rs, and thus getting rid of the rusoto import from main.rs and the long type name.